README
squid_nufw_helper is an external helper for the Squid external ACL scheme.
It provides Sigle Sign On (SSO) based on the NuFW firewall suite.

Usage
=====

The usage for the program is ::

    squid_nufw_helper -f <configuration_file> -u -c 1000

Options:
 * -f points to the config file
 * -u means squid_nufw_helper should return the found user ID
 * -c <number> tells squid_nufw_helper to process that many connexions
   before dying (and having Squid spawn a new helper process)
 * -a passes the source port in htons() to resolve byte ordering problems
   that can appear. -a is NOT needed on squid 2.6 on AMD64.


Any bug to be reported to authors, so that we improve quality of this software!
And, we are waiting for patches, too!



NEWS
2007-12-12
    * Release of squid_nufw_helper version 1.1
2007-06-12
    * Release of squid_nufw_helper version 1.0



sql.c
/* Copyright (C) 2004 Vincent Deffontaines - INL
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
* Author:  Vincent Deffontaines (vincent@inl.fr)
*
*/

#include "auth_nufw.h"

#ifdef DB_TYPE_MYSQL
const char *SQLconnect(MYSQL **sql,
                         char *host,
                         char *user,
                         char *database,
                         unsigned int port,
                         char *pass)
{
  *sql = mysql_init(*sql);
  if (*sql == NULL)
    return "NULL pointer";

/*  if (ssl_enabled)
    mysql_ssl_set(*sql,
                  ssl_key,
                  ssl_cert,
                  ssl_ca,
                  ssl_ca_dir,
                  ssl_cypher);*/

  if (! mysql_real_connect(*sql,host,user,pass,database,port,NULL,0))
//    nufw_log_error(APLOG_MARK,APLOG_WARNING,NULL,"nufw_sql_connectdb #2 : %s",mysql_error(*sql));
  {
      return (mysql_error(*sql));
  }
  return NULL;
}

int SQLlookup(SQLconnection *object,char *source_ip, char *source_port, char **user, int use_htons)
{
  MYSQL_RES *result;
  MYSQL_ROW row;
  char request[512];
  int count;
  struct in_addr src_addr;
  int port = 0;
  if (inet_pton(AF_INET,source_ip,&src_addr)<=0)
  {
      return CANT_READ_ADDRESS;
  }

  if (use_htons)
  {
      port = htons(atoi(source_port));
  }else{
      port = atoi(source_port);
  }

  if (snprintf(request,511,"SELECT DISTINCT username FROM %s WHERE (tcp_sport=%u AND ip_saddr=%lu AND (state=%d OR state=%d))",
        object->table,
        port,
        (long unsigned int)htonl(src_addr.s_addr),
        STATE_OPEN,
        STATE_ESTABLISHED)
      >=511)
      return SQL_CANT_BUILD_REQUEST;
  if (mysql_real_query(object->sql, request, strlen(request)) != 0)
      return SQL_QUERY_ERROR;

  result = mysql_store_result(object->sql);
  if (result == NULL)
      return SQL_RESULT_ERROR;

  count = mysql_num_rows(result);
  if (count < 1)
  {
      mysql_free_result(result);
      return SQL_NO_MATCH;
  }
  if (count > 1)
  {
      mysql_free_result(result);
      return SQL_TOO_MANY_MATCHES;
  }
  row = mysql_fetch_row(result);
  *user = strdup(row[0]);
  if (*user == NULL)
       return SQL_INSUFFICIENT_MEM;
  mysql_free_result(result);
  return 0;
}

int SQLclose(MYSQL *sql)
{
  if (sql != NULL)
    mysql_close(sql);

  return 0;
}
#else
char *SQLconnect(PGconn **sql,
                         char *host,
                         char *user,
                         char *database,
                         unsigned int port,
                         char *pass)
{
  char *pgsql_conninfo;
  char server_port[15];
  int pgsql_status;

  /* Guess why so many programmers write stuff for mysql rather than pg? this
   * API is not as convenient as mysql's IMHO*/
  if (snprintf(server_port,14,"%d",port) >= 14){return "snprintf() failed";}
  pgsql_conninfo = (char *)calloc(strlen(user) + strlen(pass) +
              strlen(host) + strlen(server_port) + strlen(database) +
              strlen("hostaddr='' port= dbname='' user='' password='' ") + 1,
              sizeof(char));
  if (pgsql_conninfo == NULL){return "could not calloc()";}
  strncat(pgsql_conninfo,"host='",6);
  strncat(pgsql_conninfo,host,strlen(host));
  strncat(pgsql_conninfo,"' port=",7);
  strncat(pgsql_conninfo,server_port,strlen(server_port));
  strncat(pgsql_conninfo," dbname='",9);
  strncat(pgsql_conninfo,database,strlen(database));
  strncat(pgsql_conninfo,"' user='",8);
  strncat(pgsql_conninfo,user,strlen(user));
  strncat(pgsql_conninfo,"' password='",12);
  strncat(pgsql_conninfo,pass,strlen(pass));
  strncat(pgsql_conninfo,"'",1);

  *sql = PQconnectdb(pgsql_conninfo);
  pgsql_status = PQstatus(*sql);

  free(pgsql_conninfo);
  if(pgsql_status != CONNECTION_OK) {
    PQfinish(*sql);
    return "Cannot connect to PostgreSQL database.";
  }
  return NULL;
}

int SQLclose(PGconn *sql)
{
  if (sql != NULL)
    PQfinish(sql);
  return 0;
}

int SQLlookup(SQLconnection *object,char *source_ip, char *source_port, char **user, int use_htons)
{
  PGresult *result;
  char request[512];
  int count;
  int port = 0;
  char *str_res;
  if (use_htons)
  {
      port = htons(atoi(source_port));
  }else{
      port = atoi(source_port);
  }
  if (snprintf(request,511,"SELECT DISTINCT username FROM %s WHERE (tcp_sport=%u AND ip_saddr='%s' AND (state=%d OR state=%d))",
        object->table,
        port,
        source_ip,
        STATE_OPEN,
        STATE_ESTABLISHED)
      >=511)
      return SQL_CANT_BUILD_REQUEST;

  result = PQexec(object->sql,request);
  if (PQresultStatus(result) != PGRES_COMMAND_OK)
  {
      PQclear(result);
      return SQL_QUERY_ERROR;
  }
  if (result == NULL)
  {
      PQclear(result);
      return SQL_QUERY_ERROR;
  }

  count = PQntuples(result);
  if (count < 1)
  {
      PQclear(result);
      return SQL_NO_MATCH;
  }
  if (count > 1)
  {
      PQclear(result);
      return SQL_TOO_MANY_MATCHES;
  }
  if (PQgetisnull(result,0,0))
  {
      PQclear(result);
      return SQL_NO_MATCH;
  }
  str_res = PQgetvalue(result,0,0);
  PQclear(result);
  *user = strdup(str_res);
  if (*user == NULL)
      return SQL_INSUFFICIENT_MEM;
  return 0;
}
#endif


conf.c
/* $Id: dict.c,v 1.2 2003/01/23 00:36:01 robertc Exp $
* Copyright (C) 2002 Rodrigo Campos
*           (C) 2004-2008 Vincent Deffontaines - INL
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
* Author: Rodrigo Campos (rodrigo@geekbunker.org)
* Adaptation to NuFW by Vincent Deffontaines
*
*/


#include "auth_nufw.h"

#ifndef DEBUG
#undef DEBUG
#endif

const char *const w_space = " \t\n\r";


size_t max(size_t a, size_t b)
{
  size_t res = a;
  if (b>a)
      res = b;
  return res;
}

static void
parse_char(char **ourchar)
{
  char *token = strtok(NULL,w_space);
  if (token == NULL)
      *ourchar = NULL;
  else
  {
      *ourchar = strdup(token);
  }
}

static void
parse_unsigned_int(unsigned int *value)
{
  char *token = strtok(NULL,w_space);
  if (token == NULL)
      *value = 0;
  else
  {
      *value = atoi(token);
  }
}



SQLconnection *
read_conf (FILE * FH)
{
  SQLconnection *our_conn;
  char line[BUFSIZE];		/* the buffer for the lines read
				   from the config file */
  char *cp;
  char *token;			/* a char pointer used to parse
				   each line */

  our_conn = malloc (sizeof (SQLconnection));
#ifdef DB_TYPE_MYSQL
  our_conn->port = 3306;
#else
  our_conn->port = 5432;
#endif

  while ((cp = fgets (line, sizeof (line), FH)) != NULL) {
		  if (line[0] == '#') {
				  continue;
		  }
    if ((cp = strchr (line, '\n')) != NULL) {
      /* chop \n characters */
      *cp = '\0';
    }
    if (line == NULL)
        continue;
    if ((token = strtok(line, w_space)) == NULL)
        continue; //Ignore empty lines;
    else if(!strcmp(token,"db_host"))
        parse_char(&our_conn->host);
    else if(!strcmp(token,"db_port"))
        parse_unsigned_int(&our_conn->port);
    else if(!strcmp(token,"db_database"))
        parse_char(&our_conn->database);
    else if(!strcmp(token,"db_table"))
        parse_char(&our_conn->table);
    else if(!strcmp(token,"db_user"))
        parse_char(&our_conn->user);
    else if(!strcmp(token,"db_pass"))
        parse_char(&our_conn->pass);
  }
  our_conn->sql = NULL;
  return our_conn;
}


autogen.sh
#!/bin/sh

find_tool_version()
{
    TOOL=$1
    MINVER=$2
    OKVER=$3
    NOKREG=$4

    for i in ${OKVER}; do
        if ${TOOL}${i} --version > /dev/null 2>&1; then
            echo "${TOOL}${i}"
            exit
        fi
    done
    if ${TOOL} --version > /dev/null 2>&1; then
        case "$(${TOOL} --version | sed -e '1s/[^0-9]*//' -e q)" in
            ${NOKREG})
            echo "You need ${TOOL} version (at least) ${MINVER} !" 1>&2
            exit 1;;
        esac
    else
        echo "You need ${TOOL} !" 1>&2
        exit 1
    fi
    echo "${TOOL}"
}

#-----------------------------------------------------------------------------

AUTOMAKE_MIN_VERSION="1.8"
AUTOMAKE_OK_VERSIONS="-1.10 110 -1.9 19 -1.8 18"
AUTOMAKE_NOK_REGEXP='0|0.*|1|1.[0-7]*'

AUTOCONF_MIN_VERSION="2.58"
AUTOCONF_OK_VERSIONS="258 259"
AUTOCONF_NOK_REGEXP='0|0.*|1|1.*|2|2.[0-4]*|2.5[0-7]*'

AUTOMAKE=$(find_tool_version automake "${AUTOMAKE_MIN_VERSION}" \
                                      "${AUTOMAKE_OK_VERSIONS}" \
                                      "${AUTOMAKE_NOK_REGEXP}") || exit 1
ACLOCAL=$(find_tool_version aclocal "${AUTOMAKE_MIN_VERSION}" \
                                    "${AUTOMAKE_OK_VERSIONS}" \
                                    "${AUTOMAKE_NOK_REGEXP}") || exit 1
AUTOCONF=$(find_tool_version autoconf "${AUTOCONF_MIN_VERSION}" \
                                      "${AUTOCONF_OK_VERSIONS}" \
                                      "${AUTOCONF_NOK_REGEXP}") || exit 1
AUTOHEADER=$(find_tool_version autoheader "${AUTOCONF_MIN_VERSION}" \
                                          "${AUTOCONF_OK_VERSIONS}" \
                                          "${AUTOCONF_NOK_REGEXP}") || exit 1

#-----------------------------------------------------------------------------

echo "[+] Run aclocal"
$ACLOCAL || exit $?

echo "[+] Run autoheader"
$AUTOHEADER || exit $?

echo "[+] Run autoconf"
$AUTOCONF || exit $?

echo "[+] Run automake"
$AUTOMAKE --add-missing --copy -Wno-portability || exit $?

echo
echo "Now type: ./configure"
echo "Help with: ./configure --help"



auth_nufw.h
/* $Id: ip_user.h,v 1.2 2003/01/23 00:36:01 robertc Exp $
* Copyright (C) 2002 Rodrigo Campos
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
* Author: Rodrigo Campos (rodrigo@geekbunker.org)
* Adaptation to NuFW by Vincent Deffontaines (vincent@inl.fr)
*
*/

#ifndef ALREADY_READ
#define ALREADY_READ

#define STATE_OPEN 0x1
#define STATE_ESTABLISHED 0x2

#define CANT_READ_ADDRESS -1
#define SQL_CANT_BUILD_REQUEST -2
#define SQL_QUERY_ERROR -3
#define SQL_RESULT_ERROR -4
#define SQL_NO_MATCH -5
#define SQL_TOO_MANY_MATCHES -6
#define SQL_INSUFFICIENT_MEM -7

#define BUFSIZE 1024

#include "config.h"

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>

#ifdef DB_TYPE_MYSQL
#  include <mysql/mysql.h>
#else
#  include <postgresql/libpq-fe.h>
#endif


typedef struct _SQLconnection {
#ifdef DB_TYPE_MYSQL
  MYSQL *sql;
#else
  PGconn *sql;
#endif
  char *host;
  unsigned int port;
  char *user;
  char *database;
  char *pass;
  char *table;
  int ssl_enabled;
  char *ssl_key;
  char *ssl_cert;
  char *ssl_ca;
  char *ssl_ca_dir;
  char *ssl_cypher;
//  struct _SQLconnection *next;
}SQLconnection;


SQLconnection *read_conf(FILE *FH);

#ifdef DB_TYPE_MYSQL
  const char *SQLconnect(MYSQL **sql, char *host, char *user, char *database, unsigned int port, char *pass);
  int SQLclose(MYSQL *sql);
#else
  char *SQLconnect(PGconn **sql, char *host, char *user, char *database, unsigned int port, char *pass);
  int SQLclose(PGconn *sql);
#endif
int SQLlookup(SQLconnection *object,char *source_ip, char *source_port, char **user, int use_htons);

#endif


main.c
/* $Id: main.c,v 1.3 2003/01/23 00:36:01 robertc Exp $
* Copyright (C) 2002 Rodrigo Campos
*           (C) 2004-2008 Vincent Deffontaines www.inl.fr
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
* Original author: Rodrigo Campos (rodrigo@geekbunker.org)
* Adaptation to NuFW by Vincent Deffontaines
*
*/

#include "auth_nufw.h"

void
usage (char *program_name)
{
  fprintf (stderr, "Usage:\n%s -f <configuration file>\n\t\t[-c <num>] (leave after that number of entries - default 1000)\n\t\t[-u] (tell squid about user when identified)\n\t\t[-a] (run htons() on source port)\n\t\t[-U] (In case of problems connecting or requesting the database, let traffic pass through unauthenticated)\n",
          program_name);
}

void check_params(SQLconnection *object)
{
  int error = 0;
  if (object->host == NULL)
  {
      fprintf(stderr,"No host specified! Use db_host\n");
      error++;
  }
  if (object->user == NULL)
  {
      fprintf(stderr,"No user specified! Use db_user\n");
      error++;
  }
  if (object->pass == NULL)
  {
      fprintf(stderr,"No password specified! Use db_pass\n");
      error++;
  }
  if (object->database == NULL)
  {
      fprintf(stderr,"No database name specified! Use db_database\n");
      error++;
  }
  if (object->table == NULL)
  {
      fprintf(stderr,"No table name specified! Use db_table\n");
      error++;
  }
  if (error > 0)
  {
      printf("Sorry. Cannot run. Please refer to the above message(s)\n");
      exit(-1);
  }
}

int
main (int argc, char *argv[])
{
  FILE *FH;
  char *filename = NULL;
  int leave_count = 1000;
  int mention_user=0;
  int ignore_sql_problems=0;
  char *program_name = argv[0];
  char *cp;
  char *source_port;
  char *source_ip;
  char line[BUFSIZE];
  int ch;
  int count;
  int use_htons = 0;
  SQLconnection *SQLconn;
#ifdef DB_TYPE_MYSQL
  const char *result;
#else
  char *result;
#endif

  setvbuf (stdout, NULL, _IOLBF, 0);
  while ((ch = getopt (argc, argv, "auUc:f:")) != -1) {
    switch (ch) {
    case 'f':
      filename = optarg;
      break;
    case 'c':
      leave_count = atoi(optarg);
      break;
    case 'a':
      use_htons = 1;
      break;
    case 'u':
      mention_user = 1;
      break;
    case 'U':
      ignore_sql_problems = 1;
      break;
    default:
      usage (program_name);
      exit (1);
    }
  }
  if (filename == NULL) {
    usage (program_name);
    exit(1);
  }
  FH = fopen (filename, "r");
  if (!FH) {
    fprintf(stderr, "Unable to open configuration file: %s\n", filename);
    exit(1);
  }
  SQLconn = read_conf(FH);
  fclose(FH);
  check_params(SQLconn);
  result = SQLconnect(&SQLconn->sql,
                            SQLconn->host,
                            SQLconn->user,
                            SQLconn->database,
                            SQLconn->port,
                            SQLconn->pass);
  if (result!=NULL)
  {
      fprintf(stderr, "helper: unable to connect to database : %s\n",result);
      if (ignore_sql_problems == 0)
      {
      	exit(1);
      }
  }

  while (fgets (line, sizeof (line), stdin)) {
    char *user = NULL;
    leave_count --;
    if (leave_count <= 0)
    {
	if (result==NULL) //only skipped if ignore_sql_problems == 1
	{
        	SQLclose(SQLconn->sql);
	}
        return 0;
    }
    if ((cp = strchr (line, '\n')) != NULL) {
      *cp = '\0';
    }
    if ((cp = strtok (line, " \t")) != NULL) {
      source_ip = cp;
      source_port = strtok (NULL, " \t");
    } else {
      fprintf (stderr, "helper: unable to read tokens\n");
      printf ("ERR\n");
      continue;
    }
    if ((source_ip == NULL) || (source_port == NULL))
    {
        fprintf (stderr, "helper: unable to read tokens %d\n",leave_count);
        printf ("ERR\n");
        continue;
    }

    if (result==NULL)
    {
     if (SQLlookup(SQLconn,source_ip,source_port,&user,use_htons) == 0) {
      printf ("OK");
      if (mention_user)
      {
          printf(" user=%s\n",user);
          free(user);
      }
      else
          printf("\n");
     } else {
      printf ("ERR\n");
      count++;
     }
    }else{
	if (ignore_sql_problems == 1) //Actually always true when reaching this code
	{
	    printf("OK\n");
	}else{
	    fprintf(stderr, "BUG : consistency options problem, please contact authors!\n");
	    exit(2);
	}
    }
  }
  if (result==NULL)
  {
  	SQLclose(SQLconn->sql);
  }
  return 0;
}
